### PR TITLE
OIDC Integration

### PR DESCRIPTION
**Pull request by carmithersh**
- [ ] All [tests](https://github.com/jfrog/frogbot#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
- [x] Update [documentation](https://github.com/jfrog/documentation) about new features / new supported technologies
---

### CODE REVIEW

File: action/src/utils.ts [hunk 1]
```diff
@@ -167,4 +174,92 @@ export class Utils {
     public static isWindows() {
         return platform().startsWith('win');
     }
+    public static async validatePlatfornUrl(jfrogUrl: string): Promise<void> {
```
Comment (hadarshjfrog) [inline action/src/utils.ts:177]: typo in "Platform"

File: action/lib/utils.js [hunk 2]
```diff
@@ -200,9 +201,93 @@ class Utils {
...
 }
 exports.Utils = Utils;
 Utils.LATEST_RELEASE_VERSION = '[RELEASE]';
 Utils.LATEST_CLI_VERSION_ARG = 'latest';
 Utils.VERSION_ARG = 'version';
 Utils.TOOL_NAME = 'frogbot';
+// OpenID Connect audience input
+Utils.OIDC_AUDIENCE_ARG = 'oidc-audience';
+// OpenID Connect provider_name input
+Utils.OIDC_INTEGRATION_PROVIDER_NAME = 'oidc-provider-name';
```
Comment (hadarshjfrog) [inline action/lib/utils.js:293]: I suggest being consistent with the naming convention - if it's an argument name - so both the OIDC args should end with _ARG

File: action/src/utils.ts [hunk 3]
```diff
@@ -6,12 +6,18 @@ import { chmodSync } from 'fs';
...
 
 export class Utils {
     private static readonly LATEST_RELEASE_VERSION: string = '[RELEASE]';
     private static readonly LATEST_CLI_VERSION_ARG: string = 'latest';
     private static readonly VERSION_ARG: string = 'version';
     private static readonly TOOL_NAME: string = 'frogbot';
+    // OpenID Connect audience input
+    private static readonly OIDC_AUDIENCE_ARG: string = 'oidc-audience';
+    // OpenID Connect provider_name input
+    private static readonly OIDC_INTEGRATION_PROVIDER_NAME: string = 'oidc-provider-name';
```
Comment (hadarshjfrog) [inline action/src/utils.ts:20]: I suggest being consistent with the naming convention - if it's an argument name - so both the OIDC args should end with _ARG

File: action/src/main.ts [hunk 4]
```diff
@@ -1,9 +1,18 @@
 import * as core from '@actions/core';
 import { Utils } from './utils';
 
+
 async function main() {
     try {
         core.startGroup('Frogbot');
+        //validate JF_URL
+        let jfrogUrl: string = process.env.JF_URL?? '';
+        if (!jfrogUrl) {
```
Comment (hadarshjfrog) [inline action/src/main.ts:10]: obviously can be in both - but if the following function is validatePlatformUrl - shouldn't we validate it there?

**Comment by hadarshjfrog** [comment]
Please add a oidc-test action, similar to the existing one in the cli

File: action/src/utils.ts [hunk 5]
```diff
@@ -37,6 +43,7 @@ export class Utils {
         await this.cacheAndAddPath(downloadDir, version, fileName);
     }
 
+
```
Comment (yahavi) [inline action/src/utils.ts:46]:
```suggestion
```

File: action/src/utils.ts [hunk 6]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
     }
+    public static async validatePlatfornUrl(jfrogUrl: string): Promise<void> {
+        const jfrogUrlFailure: string = 'JF_URL must point on your full platform URL, for example: https://mycompany.jfrog.io/, make sure the platform is up and running and accessible.'
+        //verify that the provided JFrog URL is valid and responsive
+        const pingUrl: string = jfrogUrl!.replace(/\/$/, '') + '/artifactory/api/system/ping';
+        const httpClient: HttpClient = new HttpClient();
+        let response: HttpClientResponse
+        try {
+             response = await httpClient.get(pingUrl);
+        }catch (error: any) {
```
Comment (yahavi) [inline action/src/utils.ts:185]: Please make sure to run `npm run format` to ensure no formatting issues
```suggestion
        } catch (error: any) {
```

File: action/src/utils.ts [hunk 7]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+        }
+        if (response.message.statusCode == 200) {
+            const body: string = await response.readBody();
+            if (body == 'OK') {
+                return ;
+            }else
+                throw new Error(jfrogUrlFailure);
+        } else {
+           throw new Error(jfrogUrlFailure);
+        }
```
Comment (yahavi) [inline action/src/utils.ts:196]: This code can be arrange and simplified:
```suggestion
        if (response.message.statusCode == 200) {
            const body: string = await response.readBody();
            if (body == 'OK') {
                return;
            }
        }
        throw new Error(jfrogUrlFailure);
```

File: action/src/utils.ts [hunk 8]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+    }
+
+    public static async getJfrogOIDCCredentials(jfrogUrl: string): Promise<void> {
+
```
Comment (yahavi) [inline action/src/utils.ts:200]: Redundant new line
```suggestion
```

File: action/src/utils.ts [hunk 9]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+            if (body == 'OK') {
+                return ;
+            }else
+                throw new Error(jfrogUrlFailure);
+        } else {
+           throw new Error(jfrogUrlFailure);
+        }
+    }
+
+    public static async getJfrogOIDCCredentials(jfrogUrl: string): Promise<void> {
```
Comment (yahavi) [inline action/src/utils.ts:199]: The function's name doesn't really match what it does - It doesn't return OIDC credentials from JFrog (the return value is void).

This function retrieves a JWT from GitHub and forwards it to the JFrog platform to obtain a valid access token. This process occurs only if the user hasn't supplied the `oidc-provider-name` argument.
I'd call it `setupOIDCTokenIfNeeded`.

File: action/src/utils.ts [hunk 10]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+
+        const oidcProviderName: string = core.getInput(Utils.OIDC_INTEGRATION_PROVIDER_NAME);
+        if (!oidcProviderName) {
+            // no token is set if no oidc provider was configured
```
Comment (yahavi) [inline action/src/utils.ts:203]:
```suggestion
            // No token is set if no OIDC provider was configured
```

File: action/src/utils.ts [hunk 11]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+
+
+    private static async initJfrogAccessTokenThroughOidcProtocol(
+        jfrogUrl: string,
+        jsonWebToken: string,
+        oidcProviderName: string,
+    ): Promise<void>     {
+        // assuming in this method that add parameters were provided
+
+        // If we've reached this stage, the jfrogCredentials.jfrogUrl field should hold a non-empty value obtained from process.env.JF_URL
```
Comment (yahavi) [inline action/src/utils.ts:231]: This explanation should be as JSDoc above this function. 
See, as a reference, the documentation of cacheAndAddPath.

File: action/src/utils.ts [hunk 12]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+        const responseString: string = await response.readBody();
+        const responseJson: TokenExchangeResponseData = JSON.parse(responseString);
+        process.env.JF_ACCESS_TOKEN = responseJson.access_token;
+        if (responseJson.access_token) {
+            core.setSecret(responseJson.access_token);
+        }
+        if (responseJson.errors) {
+            throw new Error(`${JSON.stringify(responseJson.errors)}`);
+        }
+        return ;
```
Comment (yahavi) [inline action/src/utils.ts:258]:
```suggestion
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Well done, @carmithersh!
Please consider my inline comments.

File: package-lock.json
Comment (yahavi) [inline package-lock.json:1]: This file should be deleted

File: testdata/projects/gradle/.gradle/7.6/checksums/checksums.lock
Comment (yahavi) [inline testdata/projects/gradle/.gradle/7.6/checksums/checksums.lock:1]: Please delete all of this directory

File: .gitignore [hunk 13]
```diff
@@ -1,7 +1,6 @@
 __tests__/runner/*
 frogbot
 frogbot.exe
-*mock*
```
[the remark this answers is not in the document]
Comment (carmithersh) [reply .gitignore:4]: its required for the http client dependency, for an inner library.

[the remark this answers is not in the document]
Comment (eranturgeman) [reply .gitignore:4]: is there any way to overpass it?
What is the required dependency that you need to push mocks for?

File: .github/workflows/action-test.yml [hunk 14]
```diff
@@ -1,9 +1,15 @@
 name: "GitHub Action Test"
+run-name: ${{ github.actor }} is testing OIDC-Access integration
```
Comment (eranturgeman) [inline .github/workflows/action-test.yml:2]:
```suggestion

```

Why do we need this?
This CI file is not intended to test the OIDC flow only.

File: .gitignore (hunk 13 above)
[the remark this answers is not in the document]
Comment (carmithersh) [reply .gitignore:4]: @action/httpclient requires undici which requires mock 
so, please add an .gitignore to all the mock files you wish excluded from git

File: action/action.yml [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    required: false
+  oidc-audience:
+    description: "By default, this is the URL of the GitHub repository owner, such as the organization that owns the repository."
+    required: false
+runs:
+  using: "node16"
+  main: "lib/main.js"
+branding:
+  icon: "terminal"
+  color: "green"
```
Comment (eranturgeman) [inline action/action.yml:20]:
```suggestion

```
This file is a duplication to action.yml that is found in the root.

File: action/src/main.ts [hunk 16]
```diff
@@ -1,9 +1,18 @@
 import * as core from '@actions/core';
 import { Utils } from './utils';
 
+
```
Comment (eranturgeman) [inline action/src/main.ts:4]:
```suggestion
```

File: action/src/utils.ts [hunk 17]
```diff
@@ -167,4 +174,92 @@ export class Utils {
     public static isWindows() {
         return platform().startsWith('win');
     }
+    public static async validatePlatformUrl(jfrogUrl: string): Promise<void> {
+        const jfrogUrlFailure: string = 'JF_URL must point on your full platform URL, for example: https://mycompany.jfrog.io/, make sure the platform is up and running and accessible.'
```
Comment (eranturgeman) [inline action/src/utils.ts:178]: If possible please define this const error message outside the function and the top of the file

File: action/src/utils.ts [hunk 18]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+        //verify that the provided JFrog URL is valid and responsive
```
Comment (eranturgeman) [inline action/src/utils.ts:179]:
```suggestion
        // Verify that the provided JFrog URL is valid and responsive
```

File: action/src/utils.ts [hunk 19]
```diff
@@ -167,4 +174,92 @@ export class Utils {
...
+        }
+    }
+
+
+    private static async initJfrogAccessTokenThroughOidcProtocol(
+        jfrogUrl: string,
+        jsonWebToken: string,
+        oidcProviderName: string,
+    ): Promise<void>     {
+        // assuming in this method that add parameters were provided
```
Comment (eranturgeman) [inline action/src/utils.ts:229]:
```suggestion
        // Assuming in this method that add parameters were provided
```
Also, what is "add parameters"?

File: testdata/projects/gradle/.gradle/buildOutputCleanup/cache.properties [hunk 20]
```diff
@@ -0,0 +1,2 @@
+#Wed May 08 15:41:54 IDT 2024
+gradle.version=7.6
```
Comment (eranturgeman) [inline testdata/projects/gradle/.gradle/buildOutputCleanup/cache.properties:2]: what are those files that were added in testdata/projectsgradle/.gradle dir?
I didn't see where you use a gradle project in this PR. Please just make sure those files were not added by mistake

File: action/src/main.ts [hunk 21]
```diff
@@ -4,6 +4,14 @@ import { Utils } from './utils';
 async function main() {
     try {
         core.startGroup('Frogbot');
+        //validate JF_URL
```
Comment (eranturgeman) [inline action/src/main.ts:7]:
```suggestion
        // Validate JF_URL
```

**Pull request closed without merging**